### ISSUE TITLE
Auto-wiring sub-property PartialBuilders

### ISSUE THREAD

**Issue by randomeizer** [JosephDuffy/Partial#145]
**Is your feature request related to a problem? Please describe.**

Yet to actually use the library, but reading through the documentation for `PartialBuilder` in the "Building complex types" section it occurred to me that it should be possible to improve this:

```swift
struct Root {
    let size1: CGSize
    let size2: CGSize
}

let rootBuilder = PartialBuilder<Root>()
let size1Builder = PartialBuilder<CGSize>()
let size2Builder = PartialBuilder<CGSize>()

let size1Subscription = size1Builder.subscribeToAllChanges { _, size1Builder in
    try? rootBuilder.setValue(size1Builder, for: \.size1)
}
let size2Subscription = size2Builder.subscribeToAllChanges { _, size2Builder in
    try? rootBuilder.setValue(size2Builder, for: \.size2)
}
```

**Describe the solution you'd like**

It should be possible to build some infrastructure to be able to write something like this instead:

```swift
struct Root {
    let size1: CGSize
    let size2: CGSize
}

let rootBuilder = PartialBuilder<Root>()
let size1Builder = rootBuilder.builder(for: \.size1)
let size2Builder = rootBuilder.builder(for: \.size2)
```

Internally, the `builder(for:)` method would wire up the subscription for you.

Potential downside is that you can't 'unsubscribe' from changes, but I'm not sure how often that would be an issue? I'm guessing that most Partial* values just get disposed once the real value is built anyway. 

A couple of options for this: 

1. Have a `SubPartialBuilder` (terrible class name, sorry) with an extra method for unsubscribing/etc if that's an issue? Eg:

```swift
// assuming that `builder(for:)` returns a `SubPartialBuilder`...
size2Builder.cancelSubscription() // or `.detach()` ?
```

Alternately to avoid needing `SubPartialBuilder`, this method could get added to the `PartialBuilder` interface and simply does nothing for top-level builders. Not sure if I like that.

2. `PartialBuilder` keeps track of the subscriptions it creates via `builder(for:)` internally, and provides a method to `unwatch` a given keypath. Eg:

```swift
rootBuilder:cancelSubscriptions(for: \.size2)
```

The upside is, you don't need any additional classes or protocols. The downside is, you can't cancel the specific sub-PartialBuilder.

On balance, I prefer option 1. And thinking about it, I'd lean towards adding `detach()` to `PartialBuilder`, rather than adding a `SubPartialBuilder` subclass. But I could see arguments for both sides.

**Describe alternatives you've considered**

The original suggestion works, just requires more work from developers, increasing the possiblity of missing steps.

**Comment by randomeizer** [comment]
I've put a pull request on two different implementations, for your reference. I prefer option 1 (#146) personally. And I might even suggest not bothering with `detach()` at all. If people want detachable `PropertyBuilder` values for some reason, they can use the manual method.

What are your thoughts?

**Comment by JosephDuffy** [comment]
Thank you for the issue and the PRs, this looks like a great change! It greatly simplifies the more complex example.

I agree that the first solution (#146) looks to be the best.

I like that you've used the existing subscriptions it's a very clean implementation.

I will take a look at why CI isn't running properly for your PR but once that's done and it's merged I'll trigger a release.

---

Sorry for the delay responding to you, somehow I wasn't watching my own repo! 🤦‍♂️

**Comment by randomeizer** [comment]
No worries. Looks like I might need to merge other changes into the branch anyway. I'll take a look shortly.

**Comment by randomeizer** [comment]
I've merged your latest changes and pushed them into both branches, just FYI.
